unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1299 @@
 }
 
-impl Task {
-    pub fn create_task(transaction: SanitizedTransaction, index: usize) -> Self {
-        Task { transaction, index }
-    }
-
+impl TaskInner {
     pub fn task_index(&self) -> usize {
         self.index
     }
 
     pub fn transaction(&self) -> &SanitizedTransaction {
         &self.transaction
     }
+
+    fn lock_attempts(&self) -> &[LockAttempt] {
+        &self.lock_attempts
+    }
+
+    fn blocked_usage_count_mut<'t>(
+        &self,
+        token: &'t mut BlockedUsageCountToken,
+    ) -> &'t mut ShortCounter {
+        self.blocked_usage_count.borrow_mut(token)
+    }
+
+    fn set_blocked_usage_count(&self, token: &mut BlockedUsageCountToken, count: ShortCounter) {
+        *self.blocked_usage_count_mut(token) = count;
+    }
+
+    #[must_use]
+    fn try_unblock(self: Task, token: &mut BlockedUsageCountToken) -> Option<Task> {
+        self.blocked_usage_count_mut(token)
+            .decrement_self()
+            .is_zero()
+            .then_some(self)
+    }
+}
+
+/// [`Task`]'s per-address attempt to use a [usage_queue](UsageQueue) with [certain kind of
+/// request](RequestedUsage).
+#[derive(Debug)]
+struct LockAttempt {
+    usage_queue: UsageQueue,
+    requested_usage: RequestedUsage,
+}
+const_assert_eq!(mem::size_of::<LockAttempt>(), 16);
+
+impl LockAttempt {
+    fn new(usage_queue: UsageQueue, requested_usage: RequestedUsage) -> Self {
+        Self {
+            usage_queue,
+            requested_usage,
+        }
+    }
+
+    fn usage_queue_mut<'t>(
+        &self,
+        usage_queue_token: &'t mut UsageQueueToken,
+    ) -> &'t mut UsageQueueInner {
+        self.usage_queue.0.borrow_mut(usage_queue_token)
+    }
+}
+
+/// Status about how the [`UsageQueue`] is used currently. Unlike [`RequestedUsage`], it has
+/// additional variant of [`Unused`](`Usage::Unused`).
+#[derive(Copy, Clone, Debug, Default)]
+enum Usage {
+    #[default]
+    Unused,
+    Readonly(ShortCounter),
+    Writable,
+}
+const_assert_eq!(mem::size_of::<Usage>(), 8);
+
+impl From<RequestedUsage> for Usage {
+    fn from(requested_usage: RequestedUsage) -> Self {
+        match requested_usage {
+            RequestedUsage::Readonly => Usage::Readonly(ShortCounter::one()),
+            RequestedUsage::Writable => Usage::Writable,
+        }
+    }
+}
+
+/// Status about how a task is requesting to use a particular [`UsageQueue`]. Unlike [`Usage`], it
+/// has only two unit variants.
+#[derive(Clone, Copy, Debug)]
+enum RequestedUsage {
+    Readonly,
+    Writable,
+}
+
+/// Internal scheduling data about a particular address.
+///
+/// Specifically, it holds the current [`Usage`] (or no usage with [`Usage::Unused`]) and which
+/// [`Task`]s are blocked to be executed after the current task is notified to be finished via
+/// [`::deschedule_task`](`SchedulingStateMachine::deschedule_task`)
+#[derive(Debug)]
+struct UsageQueueInner {
+    current_usage: Usage,
+    blocked_usages_from_tasks: VecDeque<UsageFromTask>,
```

Comment:
> In general I find UsageQueue* a bit confusing, since it puts the emphasis on a
> secondary thing. To me the main thing is that the queue holds tasks that are
> waiting. Then which access mode those tasks require is secondary.

hmm, i understand your reasoning. however, seems that there's some disagreement here. Maybe, it's just that I'm very reluctant to say something as `task`, which is actually a tuple of `(Task, <readonly or writable>)`. And treating access mode (or access request in my cognition) as primary thing isn't so bothering to me.

> Alternatively the more I read the code, the more I think UsageQueue* could be
called Resource

hehe, i and andrew actually had tangent discussion. Previously, `UsageQueue` was called `Page`.: https://github.com/solana-labs/solana/pull/35286#discussion_r1501284792

I think `Resource` is too generic in similar vein. i want to express how things are actually used or implemented with names.

> Then you could have lock_task_resources etc.

as i wrote here https://github.com/solana-labs/solana/pull/35286#discussion_r1528758840, `{lock/unlock}_task_usages` isn't super confusing to me. admittedly, I'm in search of more good names to replace `usage`, though.

Reply:
Wish I could go back and vote for LockQueue :P 

Personally I really dislike usage, and lock/unlock_task_usages doesn't sound like proper english to me. That said it sounds like names aren't set in stone yet, so we can get the logic in and continue the naming conversation at a later stage I guess.

Reply:
how about replacing `UsageQueue` with `AddressLock` and rename things all?